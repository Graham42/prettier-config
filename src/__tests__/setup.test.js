//@ts-check

// ORIG_DIR=$(pwd)
// function finish {
//     cd "$ORIG_DIR"
//     echo "Results in $TEMP_DIR"
// }
// trap finish EXIT

// TEMP_DIR="$(mktemp -d)"
// cd "$TEMP_DIR"
// npm init -y
// npx "$ORIG_DIR"
// ls -a

// echo "'.prettierrc.js' should exist"
// [ -e ".prettierrc.js" ]

// echo "'.gitignore' should exist"
// [ -e ".gitignore" ]

// echo "check:format script should exist"
// npm run check:format

// echo "fix:format script should exist"
// npm run fix:format

const fsPromises = require("fs/promises");
const os = require("os");
const path = require("path");
const process = require("process");
const { execSync } = require("child_process");

/** @type {import("tree-node-cli").default} */
//@ts-ignore the default export is bugged for the package currently
const tree = require("tree-node-cli");

const { setupPrettierConfig } = require("../setup");

describe("setup tests", () => {
  /** @type {string} */
  let workingDir;
  /** @type {string} */
  let originalCwd;
  beforeEach(async () => {
    // create a temporary directory for each test for isolation
    workingDir = await fsPromises.mkdtemp(
      path.join(os.tmpdir(), "prettier-config-test"),
    );
    originalCwd = process.cwd();
    process.chdir(workingDir);
    execSync("npm init -y", { encoding: "utf-8" });
  });
  afterEach(async () => {
    process.chdir(originalCwd);
    await fsPromises.rm(workingDir, { recursive: true, force: true });
  });

  it("should setup current version in fresh project", async () => {
    setupPrettierConfig();

    expectProjectStructure(workingDir);
    await expectProjectContents(workingDir);
  });
  it("should migrate from v1", async () => {
    execSync("npx @graham42/prettier-config@^1", { encoding: "utf-8" });
    setupPrettierConfig();

    expectProjectStructure(workingDir);
    await expectProjectContents(workingDir);
    // TODO BUG dependency should be uninstalled
  }, 20000);
  it("should migrate from v2", async () => {
    execSync("npx @graham42/prettier-config@^2", { encoding: "utf-8" });
    setupPrettierConfig();

    expectProjectStructure(workingDir);
    await expectProjectContents(workingDir);
  }, 20000);
  it("should setup current version in existing project", async () => {
    // make some modifications to assert that specific files aren't overwritten
    // run the setup
    // verify existing bits are still there
    console.log("TODO stub");
  });
});

/** @param {string} workingDir */
function expectProjectStructure(workingDir) {
  let structure = tree(workingDir, {
    allFiles: true,
    exclude: [/node_modules/],
    dirsFirst: true,
    trailingSlash: true,
  });
  structure = structure.replace(path.basename(workingDir), "__WORK_DIR__");
  expect(structure).toMatchSnapshot();
}

/**
 * Verify the contents of files created or modified by the setup
 *
 * @param {string} workingDir
 */
async function expectProjectContents(workingDir) {
  let vscodeSettingsRaw = await fsPromises.readFile(
    path.join(workingDir, ".vscode", "settings.json"),
    { encoding: "utf-8" },
  );
  expect(vscodeSettingsRaw).toMatchSnapshot();

  let vscodeExtensionsRaw = await fsPromises.readFile(
    path.join(workingDir, ".vscode", "extensions.json"),
    { encoding: "utf-8" },
  );
  expect(vscodeExtensionsRaw).toMatchSnapshot();

  let prettierConfigRaw = await fsPromises.readFile(
    path.join(workingDir, ".prettierrc.js"),
    { encoding: "utf-8" },
  );
  expect(prettierConfigRaw).toMatchSnapshot();

  let packageJsonRaw = await fsPromises.readFile(
    path.join(workingDir, "package.json"),
    { encoding: "utf-8" },
  );
  let packageJson = JSON.parse(packageJsonRaw);
  // filter out fields that can change without impact to prettier
  let relevant = {
    scripts: packageJson.scripts,
    devDependencies: packageJson.devDependencies,
    dependencies: packageJson.dependencies,
    prettier: packageJson.prettier,
  };
  expect(relevant).toMatchSnapshot();
}
