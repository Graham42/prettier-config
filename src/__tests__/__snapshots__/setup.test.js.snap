// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setup tests should migrate from v1 1`] = `
"__WORK_DIR__/
├── .vscode/
│   ├── extensions.json
│   └── settings.json
├── .gitignore
├── .prettierrc.js
├── package-lock.json
└── package.json"
`;

exports[`setup tests should migrate from v1 2`] = `
"{
  \\"[javascript]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[javascriptreact]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[typescript]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[typescriptreact]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[html]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[vue]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[css]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[less]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[scss]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[graphql]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[yaml]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[json]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[json5]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[markdown]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[mdx]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"editor.formatOnSave\\": true,
  \\"files.trimTrailingWhitespace\\": true,
  \\"files.insertFinalNewline\\": true
}
"
`;

exports[`setup tests should migrate from v1 3`] = `
"{
  \\"recommendations\\": [\\"esbenp.prettier-vscode\\"]
}
"
`;

exports[`setup tests should migrate from v1 4`] = `
"// This file and the npm scripts 'check:format' and 'fix:format' were
// generated with the command 'npx @graham42/prettier-config@latest'.
// To receive new updates, run the command again.

/**
 * @type { import(\\"prettier\\").Options }
 */
module.exports = {
  /*
   * Requiring trailing commas makes git history work better. If you have an
   * object:
   * {
   *   a: 1,
   *   b: 2
   * }
   * and then add a property after 'b':
   * {
   *   a: 1,
   *   b: 2,
   *   c: 3
   * }
   * the line with 'b' gets changed (it now requires a comma) even though
   * there's no meaningful change to it.
   *
   * When using the \`git blame\` command, git looks at line by line changes.
   * These comma-only changes make it more difficult to follow history.
   */
  trailingComma: \\"all\\",
  /**
   * When viewing a markdown file with less, cat, vim, etc. the document is
   * easier to read if the text is hard wrapped at 80 characters.
   *
   * This is a tradeoff because changes will often result in git diffs that
   * have changes across multiple lines. Fortunately GitHub has a great feature
   * for rendering differences in prose documents:
   * https://docs.github.com/en/free-pro-team@latest/github/managing-files-in-a-repository/rendering-differences-in-prose-documents
   *
   * You can also view a reasonable diff in the terminal with \`--word-diff\`.
   *
   * In the end, this is a personal preference and you're welcome to change it!
   */
  proseWrap: \\"always\\",
};
"
`;

exports[`setup tests should migrate from v1 5`] = `
Object {
  "dependencies": undefined,
  "devDependencies": Object {
    "prettier": "^2.x.x",
  },
  "prettier": undefined,
  "scripts": Object {
    "check:format": "npm run prettier -- --check",
    "fix:format": "npm run prettier -- --write",
    "prettier": "prettier --ignore-path .gitignore \\"**/*.{js,jsx,ts,tsx,html,vue,css,less,scss,graphql,yaml,yml,json,json5,md,mdx}\\"",
    "test": "echo \\"Error: no test specified\\" && exit 1",
  },
}
`;

exports[`setup tests should migrate from v2 1`] = `
"__WORK_DIR__/
├── .vscode/
│   ├── extensions.json
│   └── settings.json
├── .gitignore
├── .prettierrc.js
├── package-lock.json
└── package.json"
`;

exports[`setup tests should migrate from v2 2`] = `
"{
  \\"[javascript]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[javascriptreact]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[typescript]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[typescriptreact]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[html]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[vue]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[css]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[less]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[scss]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[graphql]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[yaml]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[json]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[json5]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[markdown]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[mdx]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"editor.formatOnSave\\": true,
  \\"files.trimTrailingWhitespace\\": true,
  \\"files.insertFinalNewline\\": true
}
"
`;

exports[`setup tests should migrate from v2 3`] = `
"{
  \\"recommendations\\": [\\"esbenp.prettier-vscode\\"]
}
"
`;

exports[`setup tests should migrate from v2 4`] = `
"// This file and the npm scripts 'check:format' and 'fix:format' were
// generated with the command 'npx @graham42/prettier-config@latest'.
// To receive new updates, run the command again.

/**
 * @type { import(\\"prettier\\").Options }
 */
module.exports = {
  /*
   * Requiring trailing commas makes git history work better. If you have an
   * object:
   * {
   *   a: 1,
   *   b: 2
   * }
   * and then add a property after 'b':
   * {
   *   a: 1,
   *   b: 2,
   *   c: 3
   * }
   * the line with 'b' gets changed (it now requires a comma) even though
   * there's no meaningful change to it.
   *
   * When using the \`git blame\` command, git looks at line by line changes.
   * These comma-only changes make it more difficult to follow history.
   */
  trailingComma: \\"all\\",
  /**
   * When viewing a markdown file with less, cat, vim, etc. the document is
   * easier to read if the text is hard wrapped at 80 characters.
   *
   * This is a tradeoff because changes will often result in git diffs that
   * have changes across multiple lines. Fortunately GitHub has a great feature
   * for rendering differences in prose documents:
   * https://docs.github.com/en/free-pro-team@latest/github/managing-files-in-a-repository/rendering-differences-in-prose-documents
   *
   * You can also view a reasonable diff in the terminal with \`--word-diff\`.
   *
   * In the end, this is a personal preference and you're welcome to change it!
   */
  proseWrap: \\"always\\",
};
"
`;

exports[`setup tests should migrate from v2 5`] = `
Object {
  "dependencies": undefined,
  "devDependencies": Object {
    "prettier": "^2.x.x",
  },
  "prettier": undefined,
  "scripts": Object {
    "check:format": "npm run prettier -- --check",
    "fix:format": "npm run prettier -- --write",
    "prettier": "prettier --ignore-path .gitignore \\"**/*.{js,jsx,ts,tsx,html,vue,css,less,scss,graphql,yaml,yml,json,json5,md,mdx}\\"",
    "test": "echo \\"Error: no test specified\\" && exit 1",
  },
}
`;

exports[`setup tests should setup current version in fresh project 1`] = `
"__WORK_DIR__/
├── .vscode/
│   ├── extensions.json
│   └── settings.json
├── .gitignore
├── .prettierrc.js
├── package-lock.json
└── package.json"
`;

exports[`setup tests should setup current version in fresh project 2`] = `
"{
  \\"[javascript]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[javascriptreact]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[typescript]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[typescriptreact]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[html]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[vue]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[css]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[less]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[scss]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[graphql]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[yaml]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[json]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[json5]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[markdown]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"[mdx]\\": { \\"editor.defaultFormatter\\": \\"esbenp.prettier-vscode\\" },
  \\"editor.formatOnSave\\": true,
  \\"files.trimTrailingWhitespace\\": true,
  \\"files.insertFinalNewline\\": true
}
"
`;

exports[`setup tests should setup current version in fresh project 3`] = `
"{
  \\"recommendations\\": [\\"esbenp.prettier-vscode\\"]
}
"
`;

exports[`setup tests should setup current version in fresh project 4`] = `
"// This file and the npm scripts 'check:format' and 'fix:format' were
// generated with the command 'npx @graham42/prettier-config@latest'.
// To receive new updates, run the command again.

/**
 * @type { import(\\"prettier\\").Options }
 */
module.exports = {
  /*
   * Requiring trailing commas makes git history work better. If you have an
   * object:
   * {
   *   a: 1,
   *   b: 2
   * }
   * and then add a property after 'b':
   * {
   *   a: 1,
   *   b: 2,
   *   c: 3
   * }
   * the line with 'b' gets changed (it now requires a comma) even though
   * there's no meaningful change to it.
   *
   * When using the \`git blame\` command, git looks at line by line changes.
   * These comma-only changes make it more difficult to follow history.
   */
  trailingComma: \\"all\\",
  /**
   * When viewing a markdown file with less, cat, vim, etc. the document is
   * easier to read if the text is hard wrapped at 80 characters.
   *
   * This is a tradeoff because changes will often result in git diffs that
   * have changes across multiple lines. Fortunately GitHub has a great feature
   * for rendering differences in prose documents:
   * https://docs.github.com/en/free-pro-team@latest/github/managing-files-in-a-repository/rendering-differences-in-prose-documents
   *
   * You can also view a reasonable diff in the terminal with \`--word-diff\`.
   *
   * In the end, this is a personal preference and you're welcome to change it!
   */
  proseWrap: \\"always\\",
};
"
`;

exports[`setup tests should setup current version in fresh project 5`] = `
Object {
  "dependencies": undefined,
  "devDependencies": Object {
    "prettier": "^2.x.x",
  },
  "prettier": undefined,
  "scripts": Object {
    "check:format": "npm run prettier -- --check",
    "fix:format": "npm run prettier -- --write",
    "prettier": "prettier --ignore-path .gitignore \\"**/*.{js,jsx,ts,tsx,html,vue,css,less,scss,graphql,yaml,yml,json,json5,md,mdx}\\"",
    "test": "echo \\"Error: no test specified\\" && exit 1",
  },
}
`;
